spec/lang/machine.md

```diff
@@ -266,12 +304,19 @@ impl<M: Memory> Machine<M> {
         ret(())
     }
 
-    pub fn terminate_active_thread(&mut self) -> NdResult {
+    fn terminate_active_thread(&mut self, frame: StackFrame<M>) -> NdResult {
         let active = self.active_thread;
         assert!(active != 0, "the main thread cannot terminate");
 
         self.threads.mutate_at(active, |thread| thread.state = ThreadState::Terminated);
 
+        // All threads that waited to join this thread get synchronized by this termination.
+        for i in ThreadId::ZERO..self.threads.len() {
+            if self.threads[i].state == ThreadState::BlockedOnJoin(active) {
```

Comment:
This function now does this loop twice -- here, and then just below to enable these threads. Please only iterate once.